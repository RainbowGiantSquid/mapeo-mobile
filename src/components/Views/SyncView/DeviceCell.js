// @flow

import React from 'react';
import {
  ActivityIndicator,
  View,
  Text,
  TouchableOpacity,
  StyleSheet,
  Dimensions
} from 'react-native';
import I18n from 'react-native-i18n';
import Icon from 'react-native-vector-icons/MaterialIcons';
import type { Device } from '../../../types/device';
import {
  DARK_BLUE,
  DARK_MANGO,
  MANGO,
  MAPEO_BLUE,
  MEDIUM_BLUE,
  VERY_LIGHT_BLUE
} from '../../../lib/styles';

type Props = {
  device: Device,
  onPress: (i: Device) => void,
  selectDevice: (device: Device) => void,
  selectedDevice?: Device,
  updateDeviceSync: (device: Device) => void
};

const styles = StyleSheet.create({
  cellContainer: {
    height: 80,
    flexDirection: 'row',
    justifyContent: 'flex-start',
    alignItems: 'center',
    borderBottomColor: DARK_BLUE,
    borderBottomWidth: 1,
    width: Dimensions.get('window').width,
    paddingHorizontal: 20
  },
  deviceName: {
    color: 'white',
    fontWeight: '700',
    fontSize: 18
  },
  deviceTextContainer: {
    flexDirection: 'column',
    justifyContent: 'center',
    alignItems: 'flex-start',
    marginLeft: 30
  },
  radioButtonOuterCircle: {
    width: 20,
    height: 20,
    backgroundColor: 'white',
    borderRadius: 50,
    justifyContent: 'center',
    alignItems: 'center',
    marginTop: -15
  },
  radioButtonInnerCircle: {
    alignSelf: 'center',
    height: 15,
    width: 15,
    borderRadius: 50,
    justifyContent: 'center',
    alignItems: 'center'
  },
  syncArrow: {
    alignSelf: 'center',
    justifyContent: 'center',
    transform: [{ rotate: '-45deg' }]
  },
  syncButtonContainer: {
    backgroundColor: 'transparent',
    position: 'absolute',
    right: 20
  },
  syncButtonInnerCircle: {
    width: 25,
    height: 25,
    borderRadius: 50,
    backgroundColor: MANGO
  },
  syncButtonOuterCircle: {
    width: 30,
    height: 30,
    borderRadius: 50,
    backgroundColor: DARK_MANGO,
    justifyContent: 'center',
    alignItems: 'center'
  },
  syncStatus: {
    color: 'white',
    fontWeight: '400',
    fontSize: 12
  }
});

I18n.fallbacks = true;
I18n.translations = {
  en: require('../../../translations/en'),
  es: require('../../../translations/es')
};

const DeviceCell = (props: Props) => {
  const handlePress = () => {
    if (props.selectedDevice) {
      props.updateDeviceSync({
        ...props.selectedDevice,
        syncStatus: 'notStarted'
      });
    }
    props.onPress(props.device);
  };
  const handleSyncUpdate = () => {
    props.updateDeviceSync({
      ...props.device,
      syncStatus: 'requested'
    });
    setTimeout(() => {
      props.updateDeviceSync({
        ...props.device,
        syncStatus: 'syncing'
      });
    }, 1000);
  };

  let syncStatusText = I18n.t('sync.via_wifi');
  if (props.device.syncStatus === 'requested') {
    syncStatusText = I18n.t('sync.requested');
  } else if (props.device.syncStatus === 'syncing') {
    syncStatusText = I18n.t('sync.syncing');
  }

  return (
    <TouchableOpacity onPress={handlePress}>
      <View
        style={[
          styles.cellContainer,
          { backgroundColor: props.device.selected ? MEDIUM_BLUE : MAPEO_BLUE }
        ]}
      >
        <View style={styles.radioButtonOuterCircle}>
          <View
            style={[
              styles.radioButtonInnerCircle,
              { backgroundColor: props.device.selected ? 'black' : MAPEO_BLUE }
            ]}
          />
        </View>
        <View style={styles.deviceTextContainer}>
          <Text style={styles.deviceName}>{props.device.name}</Text>
          <Text style={styles.syncStatus}>{syncStatusText}</Text>
        </View>
        {props.device.selected &&
          props.device.syncStatus === 'notStarted' && (
            <TouchableOpacity
              style={styles.syncButtonContainer}
              onPress={handleSyncUpdate}
            >
              <View style={styles.syncButtonOuterCircle}>
                <View style={styles.syncButtonInnerCircle}>
                  <Icon
                    color="white"
                    name="arrow-forward"
                    size={23}
                    style={styles.syncArrow}
                  />
                </View>
              </View>
            </TouchableOpacity>
          )}
        {props.device.selected &&
          props.device.syncStatus !== 'notStarted' && (
            <View style={styles.syncButtonContainer}>
              <ActivityIndicator size="large" color={VERY_LIGHT_BLUE} />
            </View>
          )}
      </View>
    </TouchableOpacity>
  );
};

export default DeviceCell;
