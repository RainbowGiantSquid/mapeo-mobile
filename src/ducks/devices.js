// @flow
import update from 'immutability-helper';
import { keyBy } from 'lodash';
import { create } from '../lib/redux';
import type { SyncStatus } from '../types/device';

export const {
  type: DEVICE_LIST,
  action: deviceList,
  reducer: deviceListReducer
} = create('DEVICE_LIST', {
  success: (state, meta, payload) => {
    const newState = update(state, { devices: { $set: keyBy(payload, 'id') } });

    return newState;
  }
});

export const {
  type: DEVICE_SELECT,
  action: deviceSelect,
  reducer: deviceSelectReducer
} = create('DEVICE_SELECT', {
  start: (state, meta: string) => {
    const newState = update(state, { selectedDevice: { $set: meta } });

    return newState;
  }
});

export type DeviceSyncUpdateMeta = {
  id: string,
  status: SyncStatus
};

export const {
  type: DEVICE_SYNC_UPDATE,
  action: deviceSyncUpdate,
  reducer: deviceSyncUpdateReducer
} = create('DEVICE_SYNC_UPDATE', {
  start: (state, meta: DeviceSyncUpdateMeta) => {
    if (!state.devices[meta.id]) {
      return state;
    }

    const newState = update(state, {
      devices: { [meta.id]: { syncStatus: { $set: meta.status } } }
    });

    return newState;
  }
});

export default [
  deviceListReducer,
  deviceSelectReducer,
  deviceSyncUpdateReducer
];
