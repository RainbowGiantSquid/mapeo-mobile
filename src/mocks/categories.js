// @flow

export const createCategory = (category?: any) => ({
  id: 'categoryID',
  name: 'categoryName',
  fieldIds: [7],

  ...category
});
