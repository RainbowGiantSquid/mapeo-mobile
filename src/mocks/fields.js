// @flow

export const createField = (field?: any) => ({
  id: 'fieldId',
  name: 'fieldName',
  type: 'text',
  placeholder: 'this is some placeholder text',
  answered: false,
  answer: '',

  ...field
});
