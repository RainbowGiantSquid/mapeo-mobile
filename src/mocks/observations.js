// @flow

export const createObservation = (observation?: any) => ({
  type: 'Observation',
  id: 'ObservationID',
  lat: 50.5,
  lon: 50.5,
  link: 'linkID',
  created: 'Thu Apr 05 2018 23:06:45 GMT+0200 (CEST)',
  name: 'Oil Spill',
  notes:
    'This is pretty bad. It smells pretty noxious, and the entire pond is affected.',
  observedBy: 'You',
  media: [],
  icon: null,

  ...observation
});
