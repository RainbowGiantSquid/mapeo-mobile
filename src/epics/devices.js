// @flow

import React from 'react';
import type { ActionsObservable } from 'redux-observable';
import { DEVICE_LIST, deviceList } from '../ducks/devices';
import { createDevice } from '../mocks/devices';
import type { Action } from '../types/redux';
import type { Device } from '../types/device';

const initialDevices = [
  {
    name: `Aldo's MacBook`,
    selected: false,
    syncStatus: 'notStarted'
  },
  {
    name: `Cindy's MacBook`,
    selected: false,
    syncStatus: 'notStarted'
  },
  {
    name: `Gregor's MacBook`,
    selected: false,
    syncStatus: 'notStarted'
  }
];

// const initialDevices = [];

export const deviceListEpic = (
  action$: ActionsObservable<Action<string, Device[]>>
) =>
  action$
    .ofType(DEVICE_LIST)
    .filter(action => action.status === 'Start')
    .map(() =>
      deviceList(
        '',
        initialDevices.map((d, i) =>
          createDevice({
            name: d.name,
            id: i.toString(),
            selected: d.selected,
            syncStatus: d.syncStatus
          })
        )
      )
    );

export default [deviceListEpic];
