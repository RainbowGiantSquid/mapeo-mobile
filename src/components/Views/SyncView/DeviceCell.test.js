// @flow
import React from 'react';
import 'react-native';
import renderer from 'react-test-renderer';
import DeviceCell from './DeviceCell';
import { createDevice } from '../../../mocks/devices';

describe('DeviceCell tests', () => {
  const device = createDevice();
  const selectedDevice = device;
  const updateDeviceSync = jest.fn();
  const onPress = jest.fn();

  test('snapshot', () => {
    const tree = renderer.create(
      <DeviceCell
        onPress={onPress}
        device={device}
        selectedDevice={selectedDevice}
        updateDeviceSync={updateDeviceSync}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
