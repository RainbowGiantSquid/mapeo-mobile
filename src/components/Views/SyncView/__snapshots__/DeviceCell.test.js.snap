// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeviceCell tests snapshot 1`] = `
<View
  accessibilityComponentType={undefined}
  accessibilityLabel={undefined}
  accessibilityTraits={undefined}
  accessible={true}
  collapsable={undefined}
  hasTVPreferredFocus={undefined}
  hitSlop={undefined}
  isTVSelectable={true}
  nativeID={undefined}
  onLayout={undefined}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID={undefined}
  tvParallaxProperties={undefined}
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "borderBottomColor": "#17317c",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "height": 80,
          "justifyContent": "flex-start",
          "paddingHorizontal": 20,
          "width": 750,
        },
        Object {
          "backgroundColor": "#3366ff",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "white",
          "borderRadius": 50,
          "height": 20,
          "justifyContent": "center",
          "marginTop": -15,
          "width": 20,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "alignSelf": "center",
              "borderRadius": 50,
              "height": 15,
              "justifyContent": "center",
              "width": 15,
            },
            Object {
              "backgroundColor": "#3366ff",
            },
          ]
        }
      />
    </View>
    <View
      style={
        Object {
          "alignItems": "flex-start",
          "flexDirection": "column",
          "justifyContent": "center",
          "marginLeft": 30,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontSize": 18,
            "fontWeight": "700",
          }
        }
      >
        deviceName
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "white",
            "fontSize": 12,
            "fontWeight": "400",
          }
        }
      >
        via WiFi
      </Text>
    </View>
  </View>
</View>
`;
